feat: update mcms lib with new timelock converter and update BuildProposalFromBatches() for solana chain metadata

ecPablo opened the pull request:
This PR changes the `BuildProposalFromBatches` to accept an `env` parameter to be able to extract role account addresses from the address book. This is necessary to build solana mcms proposal as the chain metadata for solana now requires we include the account addresses for the different mcms roles used in the timelock.

## AI Summary
This pull request includes several changes to the `deployment` package, primarily focused on simplifying the usage of the `BuildProposalFromBatchesV2` function and modifying the `McmsTimelockConverterForChain` function. The most important changes include removing unnecessary context parameters and updating the `McmsTimelockConverterForChain` function to no longer require the `env` parameter for Solana chains.

### Simplification of `BuildProposalFromBatchesV2` function calls:
* [`deployment/ccip/changeset/deployer_group.go`](diffhunk://#diff-6e11d25caefb9c749fd2eb67adf89a14e8698c5651849fd5055419ef2c76e2a9L251-R251): Removed `GetContext` method call from the `BuildProposalFromBatchesV2` function.
* [`deployment/ccip/changeset/solana/cs_deploy_chain.go`](diffhunk://#diff-f4968f319e481ae85ae26a7800debb755d306eea18580bddface5d8ea614d887L139-R139): Removed `GetContext` method call from the `BuildProposalFromBatchesV2` function.
* [`deployment/ccip/changeset/v1_6/cs_ccip_home.go`](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L439-R439): Removed `GetContext` method call from multiple functions. [[1]](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L439-R439) [[2]](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L688-R688) [[3]](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L877-R877) [[4]](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L1165-R1165) [[5]](diffhunk://#diff-3ab7b3159b26540efd60b40eba7d0ed3fdb5667d285694c850c11b17185d0962L1371-R1371)

### Changes to `McmsTimelockConverterForChain` function:
* [`deployment/common/proposalutils/mcms_helpers.go`](diffhunk://#diff-21c53969f9bc869d7b6ec247331547656e67077caf443c36c999b98088bebec2L284-R284): Updated `McmsTimelockConverterForChain` to no longer require the `env` parameter for Solana chains. [[1]](diffhunk://#diff-21c53969f9bc869d7b6ec247331547656e67077caf443c36c999b98088bebec2L284-R284) [[2]](diffhunk://#diff-21c53969f9bc869d7b6ec247331547656e67077caf443c36c999b98088bebec2L294-R294) [[3]](diffhunk://#diff-21c53969f9bc869d7b6ec247331547656e67077caf443c36c999b98088bebec2L305-R313)
* [`deployment/common/proposalutils/mcms_test_helpers.go`](diffhunk://#diff-f6d460202625a0505636ab1a63f111f146e0fcd7de84833158a0a820d1a373cbL121-R121): Updated test helpers to reflect changes in `McmsTimelockConverterForChain`.

### Commented out unused code:
* [`deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock.go`](diffhunk://#diff-1112b5e423bbda2962138a65f399667548a26ab5c9eecbcadd5e1d23eec77ad0R142): Commented out unused `additionalFieldsMetadata` code. [[1]](diffhunk://#diff-1112b5e423bbda2962138a65f399667548a26ab5c9eecbcadd5e1d23eec77ad0R142) [[2]](diffhunk://#diff-1112b5e423bbda2962138a65f399667548a26ab5c9eecbcadd5e1d23eec77ad0R152-R156)

These changes streamline the code by removing unnecessary context parameters and simplify the `McmsTimelockConverterForChain` function for Solana chains.

ecPablo committed:
feat: update mcms lib with new timelock converter and update BuildProposalFromBatchesV2() to accept env for extracting solana metadata.

ecPablo committed:
feat: update go mod to new mcms and ccip-solana version

ecPablo committed:
Merge branch 'refs/heads/develop' into ecpablo/update-mcms-chain-metadata

# Conflicts:
#	core/scripts/go.mod
#	core/scripts/go.sum
#	deployment/go.mod
#	deployment/go.sum
#	integration-tests/go.sum
#	integration-tests/load/go.sum
#	system-tests/lib/go.mod
#	system-tests/lib/go.sum
#	system-tests/tests/go.mod
#	system-tests/tests/go.sum

ecPablo committed:
feat: update go mod to new mcms and ccip-solana version

ecPablo committed:
fix: go mod

ecPablo committed:
fix: go mod

ecPablo committed:
fix: go mod core scripts

ecPablo committed:
chore: add changeset

ecPablo committed:
fix: update tx field name to RawData

graham-chainlink commented on deployment/ccip/changeset/solana/transfer_ccip_to_mcms_with_timelock.go:
```diff
@@ -148,6 +149,11 @@ func TransferCCIPToMCMSWithTimelockSolana(
 			mcmState.TimelockProgram,
 			mcmsSolana.PDASeed(mcmState.TimelockSeed),
 		)
+		//additionalFieldsMetadata[solChain.Selector], err = json.Marshal(mcmsSolana.AdditionalFieldsMetadata{
+		//	ProposerRoleAccessController:  mcmState.ProposerAccessControllerAccount,
+		//	CancellerRoleAccessController: mcmState.CancellerAccessControllerAccount,
+		//	BypasserRoleAccessController:  mcmState.BypasserAccessControllerAccount,
+		//})
```
remove comments?

graham-chainlink replied:
removed in latest commit

graham-chainlink commented on deployment/common/proposalutils/propose.go:
```diff
@@ -142,9 +146,20 @@ func BuildProposalFromBatchesV2(
-
+func getSolAccessControllerFromAddress(addresses map[string]deployment.TypeAndVersion, typeAndVersion deployment.TypeAndVersion) (solana.PublicKey, error) {
+	for addr, tv := range addresses {
+		if tv.Type == tv.Type && tv.Version == tv.Version {
```
```suggestion
		if tv.Type == ?? && tv.Version == ?? {
```

graham-chainlink replied:
fixed in latest commit

graham-chainlink committed:
fix: more cleanup

- update tests
- go mod tidy
- delete comments
- other bug fixes

gustavogama-cll commented on deployment/common/changeset/test_helpers.go:
```diff
@@ -236,3 +237,17 @@ func DeployLinkTokenTest(t *testing.T, solChains int) {
 	}
 }
+
+func SetPreloadedSolanaAddresses(t *testing.T, env deployment.Environment, selector uint64) {
```
[nit][noaction]

I'm uncomfortable with how we tend to place test helpers in the same package as application code. It's very easy for someone who's new to the code base to see this function as an option provided by his/hers IDE and think that it's a feature provided by the package.

graham-chainlink replied:
yeah unfortunately i had to move `SetPreloadedSolanaAddresses` as it seems to be the best place for now without duplication `SetPreloadedSolanaAddresses` everywhere which is what we doing now.

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -156,14 +172,56 @@ func buildProposalMetadataV2(
+				MCMAddress:      proposerMcms,
+			}
+		case chain_selectors.FamilySolana:
+			mcmsProgramID, mcmSeed, err := mcmssolanasdk.ParseContractAddress(proposerMcms)
```
Isn't it easier to use `state.MaybeLoadMCMSWithTimelockStateSolana` instead of searching for each contract type?

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -143,8 +147,20 @@ func BuildProposalFromBatchesV2(
+		}
+	}
+	return solana.PublicKey{}, errors.New("address not found")
+}
```
[nit] space before `func`

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -1,19 +1,24 @@
 	"github.com/smartcontractkit/mcms/types"
+
+	"github.com/smartcontractkit/chainlink/deployment"
+	types2 "github.com/smartcontractkit/chainlink/deployment/common/types"
```
[nit] commontypes?

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -1,19 +1,24 @@
 	mcmslib "github.com/smartcontractkit/mcms"
 	"github.com/smartcontractkit/mcms/sdk"
+	mcmssolanasdk "github.com/smartcontractkit/mcms/sdk/solana"
 	"github.com/smartcontractkit/mcms/types"
```
[nit] standardize on `mcmstypes`?

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -1,19 +1,24 @@
 	"github.com/smartcontractkit/ccip-owner-contracts/pkg/proposal/timelock"
+	chain_selectors "github.com/smartcontractkit/chain-selectors"
 	mcmslib "github.com/smartcontractkit/mcms"
 	"github.com/smartcontractkit/mcms/sdk"
```
[nit] standardize on `mcmssdk`

graham-chainlink replied:
done

gustavogama-cll commented on deployment/common/proposalutils/propose_test.go:
```diff
@@ -120,8 +160,21 @@ func TestBuildProposalFromBatchesV2(t *testing.T) {
-				assert.Equal(t, map[types.ChainSelector]types.ChainMetadata{0xc9f9284461c852b: {StartingOpCount: 0x0, MCMAddress: mcmsState.ProposerMcm.Address().String()}}, proposal.ChainMetadata)
-				assert.Equal(t, timelockAddress.String(), proposal.TimelockAddresses[types.ChainSelector(chainSelector)])
+				assert.Equal(t, map[types.ChainSelector]types.ChainMetadata{
+					0xc9f9284461c852b: {
```
maybe use `909606746561742123` as in the rest of the codebase?

graham-chainlink replied:
silly me, i can just use the variable here.

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -156,14 +172,56 @@ func buildProposalMetadataV2(
+				StartingOpCount: opCount,
+				MCMAddress:      proposerMcms,
+			}
+		case chain_selectors.FamilySolana:
+			mcmsProgramID, mcmSeed, err := mcmssolanasdk.ParseContractAddress(proposerMcms)
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse solana contract address: %w", err)
+			}
+			// Get access controller from state
+			addresses, err := env.ExistingAddresses.AddressesForChain(selector)
+			if err != nil {
+				return nil, fmt.Errorf("failed to load addresses for chain %d: %w", selector, err)
+			}
+			proposerAC, err := getSolAccessControllerFromAddress(addresses, deployment.NewTypeAndVersion(types2.ProposerAccessControllerAccount, deployment.Version1_0_0))
+			if err != nil {
+				return nil, fmt.Errorf("failed to get proposer access controller: %w", err)
+			}
+			cancellerAC, err := getSolAccessControllerFromAddress(addresses, deployment.NewTypeAndVersion(types2.CancellerAccessControllerAccount, deployment.Version1_0_0))
+			if err != nil {
+				return nil, fmt.Errorf("failed to get canceller access controller: %w", err)
+			}
+			bypasserAC, err := getSolAccessControllerFromAddress(addresses, deployment.NewTypeAndVersion(types2.BypasserAccessControllerAccount, deployment.Version1_0_0))
+			if err != nil {
+				return nil, fmt.Errorf("failed to get bypasser access controller: %w", err)
+			}
+			metadata, err := mcmssolanasdk.NewChainMetadata(
+				opCount,
+				mcmsProgramID,
+				mcmSeed,
+				proposerAC,
+				cancellerAC,
+				bypasserAC)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create chain metadata: %w", err)
+			}
+			metaDataPerChain[chainID] = metadata
+
 		}
```
```suggestion
		case chain_selectors.FamilySolana:
			addresses, err := env.ExistingAddresses.AddressesForChain(selector)
			if err != nil {
				return nil, fmt.Errorf("failed to load addresses for chain %d: %w", selector, err)
			}

			solanaState, err := state.MaybeLoadMCMSWithTimelockChainStateSolana(env.SolChains[selector], addresses)
			if err != nil {
				return nil, fmt.Errorf("failed to load solana state: %w", err)
			}

			// FIXME: maybe need to parse the seed first
			// _, mcmSeed, err := mcmssolanasdk.ParseContractAddress(solanaState.ProposerMcmSeed)
			// if err != nil {
			// 	return nil, fmt.Errorf("failed to parse solana contract address: %w", err)
			// }

			metaDataPerChain[chainID], err = mcmssolanasdk.NewChainMetadata(
				opCount,
				solanaState.McmProgram,
				mcmssolanasdk.PDASeed(solanaState.ProposerMcmSeed),
				solanaState.ProposerAccessControllerAccount,
				solanaState.CancellerAccessControllerAccount,
				solanaState.BypasserAccessControllerAccount)
			if err != nil {
				return nil, fmt.Errorf("failed to create chain metadata: %w", err)
			}
		}
```

graham-chainlink replied:
Thanks updated!

graham-chainlink replied:
Hmm, import cycle on `"github.com/smartcontractkit/chainlink/deployment/common/changeset/state"` , we would have to move something around to make this work.

graham-chainlink replied:
I reverted the change for now until we can find a better way to resolve the import cycle

gustavogama-cll replied:
If you want, you can try the following:
```go
diff --git a/deployment/common/changeset/state/evm.go b/deployment/common/changeset/state/evm.go
index 56aec84533..9f006b5477 100644
--- a/deployment/common/changeset/state/evm.go
+++ b/deployment/common/changeset/state/evm.go
@@ -8,7 +8,6 @@ import (
 	bindings "github.com/smartcontractkit/ccip-owner-contracts/pkg/gethwrappers"
 
 	"github.com/smartcontractkit/chainlink/deployment"
-	"github.com/smartcontractkit/chainlink/deployment/common/proposalutils"
 	"github.com/smartcontractkit/chainlink/deployment/common/types"
 	view "github.com/smartcontractkit/chainlink/deployment/common/view/v1_0"
 	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/link_token_interface"
@@ -20,7 +19,32 @@ import (
 // It is public for use in product specific packages.
 // Either all fields are nil or all fields are non-nil.
 type MCMSWithTimelockState struct {
-	*proposalutils.MCMSWithTimelockContracts
+	CancellerMcm *bindings.ManyChainMultiSig
+	BypasserMcm  *bindings.ManyChainMultiSig
+	ProposerMcm  *bindings.ManyChainMultiSig
+	Timelock     *bindings.RBACTimelock
+	CallProxy    *bindings.CallProxy
+}
+
+// Validate checks that all fields are non-nil, ensuring it's ready
+// for use generating views or interactions.
+func (state MCMSWithTimelockState) Validate() error {
+	if state.Timelock == nil {
+		return errors.New("timelock not found")
+	}
+	if state.CancellerMcm == nil {
+		return errors.New("canceller not found")
+	}
+	if state.ProposerMcm == nil {
+		return errors.New("proposer not found")
+	}
+	if state.BypasserMcm == nil {
+		return errors.New("bypasser not found")
+	}
+	if state.CallProxy == nil {
+		return errors.New("call proxy not found")
+	}
+	return nil
 }
 
 func (state MCMSWithTimelockState) GenerateMCMSWithTimelockView() (view.MCMSWithTimelockView, error) {
@@ -61,9 +85,8 @@ func MaybeLoadMCMSWithTimelockState(env deployment.Environment, chainSelectors [
 // - It only found part of the bundle of contracts
 // - If found more than one instance of a contract (we expect one bundle in the given addresses)
 func MaybeLoadMCMSWithTimelockChainState(chain deployment.Chain, addresses map[string]deployment.TypeAndVersion) (*MCMSWithTimelockState, error) {
-	state := MCMSWithTimelockState{
-		MCMSWithTimelockContracts: &proposalutils.MCMSWithTimelockContracts{},
-	}
+	state := MCMSWithTimelockState{}
+
 	// We expect one of each contract on the chain.
 	timelock := deployment.NewTypeAndVersion(types.RBACTimelock, deployment.Version1_0_0)
 	callProxy := deployment.NewTypeAndVersion(types.CallProxy, deployment.Version1_0_0)
diff --git a/deployment/common/changeset/state/evm_test.go b/deployment/common/changeset/state/evm_test.go
index 0a5a639ece..cd6d51346d 100644
--- a/deployment/common/changeset/state/evm_test.go
+++ b/deployment/common/changeset/state/evm_test.go
@@ -15,7 +15,6 @@ import (
 	"go.uber.org/zap/zapcore"
 
 	"github.com/smartcontractkit/chainlink/deployment"
-	"github.com/smartcontractkit/chainlink/deployment/common/proposalutils"
 	"github.com/smartcontractkit/chainlink/deployment/environment/memory"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
 )
@@ -45,14 +44,14 @@ func TestMCMSWithTimelockState_GenerateMCMSWithTimelockViewV2(t *testing.T) {
 		common.HexToAddress("0x0000000000000000000000000000000000000009"))
 
 	tests := []struct {
-		name      string
-		contracts *proposalutils.MCMSWithTimelockContracts
-		want      string
-		wantErr   string
+		name    string
+		state   *MCMSWithTimelockState
+		want    string
+		wantErr string
 	}{
 		{
 			name: "success",
-			contracts: &proposalutils.MCMSWithTimelockContracts{
+			state: &MCMSWithTimelockState{
 				ProposerMcm:  proposerMcm,
 				CancellerMcm: cancellerMcm,
 				BypasserMcm:  bypasserMcm,
@@ -110,9 +109,7 @@ func TestMCMSWithTimelockState_GenerateMCMSWithTimelockViewV2(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			state := MCMSWithTimelockState{tt.contracts}
-
-			got, err := state.GenerateMCMSWithTimelockView()
+			got, err := tt.state.GenerateMCMSWithTimelockView()
 
 			if tt.wantErr == "" {
 				require.NoError(t, err)

```

graham-chainlink replied:
Very nice @gustavogama-cll , i didn dig deep into it to find a better way but what you have here works and very minimal, committed into a new latest commit

ecPablo replied:
Nice guys thanks, using the state function looks a lot better

gustavogama-cll commented on deployment/common/proposalutils/propose.go:
```diff
@@ -143,10 +147,23 @@ func BuildProposalFromBatchesV2(
 	return build, nil
 }
 
+func getSolAccessControllerFromAddress(addresses map[string]deployment.TypeAndVersion, typeAndVersion deployment.TypeAndVersion) (solana.PublicKey, error) {
```
this is likely unused now

graham-chainlink committed:
fix: break dep cycle

for state package, stop relying on proposalutils package so the `buildProposalMetadataV2` can call `MaybeLoadMCMSWithTimelockChainStateSolana`

ecPablo committed:
Merge branch 'develop' into ecpablo/update-mcms-chain-metadata

# Conflicts:
#	core/scripts/go.mod
#	core/scripts/go.sum
#	deployment/go.mod
#	deployment/go.sum
#	integration-tests/go.sum
#	integration-tests/load/go.sum
#	system-tests/lib/go.mod
#	system-tests/lib/go.sum
#	system-tests/tests/go.mod
#	system-tests/tests/go.sum

ecPablo committed:
fix: merge conflicts develop

ecPablo committed:
fix: go sum

ecPablo committed:
fix: update lib

ecPablo committed:
fix: upgrade ccip solana

ecPablo committed:
fix: downgrade ccip solana for rest of core packages and chainlink/deployment

ecPablo committed:
Merge branch 'refs/heads/develop' into ecpablo/update-mcms-chain-metadata

# Conflicts:
#	core/scripts/go.mod
#	core/scripts/go.sum
#	deployment/go.mod
#	integration-tests/go.sum
#	integration-tests/load/go.sum
#	system-tests/lib/go.mod
#	system-tests/lib/go.sum
#	system-tests/tests/go.mod
#	system-tests/tests/go.sum

ecPablo committed:
fix: go mod core scripts add replace

ecPablo committed:
fix: downgrade solana from mcms lib

ecPablo committed:
Merge branch 'develop' into ecpablo/update-mcms-chain-metadata

# Conflicts:
#	integration-tests/go.mod
#	integration-tests/load/go.mod

ecPablo committed:
fix: merge conflicts develop

ecPablo committed:
fix: merge conflicts develop

ecPablo committed:
fix: go mod deps

ecPablo committed:
Merge branch 'develop' into ecpablo/update-mcms-chain-metadata

# Conflicts:
#	deployment/ccip/changeset/solana/cs_add_remote_chain.go

ecPablo committed:
fix: params in solana utils
